src/casts/specify-representation.rkt

```diff
@@ -443,33 +412,75 @@ but a static single assignment is implicitly maintained.
+            (assign$ t1a (sr-tagged-array-ref t1 TYPE-TUPLE-TAG i))
+            (assign$ t2a (sr-tagged-array-ref t2 TYPE-TUPLE-TAG i))
+            (assign$ val-crcn (app-code$ mk-crcn-label t1a t2a l))
+            (assign$ c-i (op$ + (op$ - i TYPE-TUPLE-ELEMENTS-OFFSET)
+                              COERCION-TUPLE-ELEMENTS-OFFSET))
+            (sr-array-set! crcn c-i val-crcn))
+          (sr-array-set! crcn COERCION-TUPLE-COUNT-INDEX tagged-count)
+          (sr-tag-value crcn COERCION-MEDIATING-TAG)]
+         [(op$ > t1-count t2-count)
+          (assign$ crcn UNIT-IMDT)
+          (assign$ id? TRUE-IMDT)
+          (assign$ iters (op$ + TYPE-TUPLE-ELEMENTS-OFFSET t2-count))
+          (repeat$ (i TYPE-TUPLE-ELEMENTS-OFFSET iters) (_ UNIT-IMDT)
+            (assign$ t1a (sr-tagged-array-ref t1 TYPE-TUPLE-TAG i))
+            (assign$ t2a (sr-tagged-array-ref t2 TYPE-TUPLE-TAG i))
+            (assign$ val-crcn (app-code$ mk-crcn-label t1a t2a l))
+            (cond$
+             [id?
+              (assign$ tmp-id?
+                (sr-check-tag=?
+                 val-crcn COERCION-TAG-MASK COERCION-IDENTITY-TAG))
+              (cond$
+               [(op$ not tmp-id?)
+                (assign$ tagged-count
+                  (op$ + (op$ %<< t1-count COERCION-SECOND-TAG-SHIFT)
+                       COERCION-TUPLE-SECOND-TAG))
+                (assign$ c-i (op$ + (op$ - i TYPE-TUPLE-ELEMENTS-OFFSET)
+                                  COERCION-TUPLE-ELEMENTS-OFFSET))
+                (Assign (Var-id crcn)
+                  (op$ Alloc (op$ + t1-count COERCION-TUPLE-ELEMENTS-OFFSET)))
+                (sr-array-set! crcn COERCION-TUPLE-COUNT-INDEX tagged-count)
+                (repeat$ (j COERCION-TUPLE-ELEMENTS-OFFSET c-i) (_ UNIT-IMDT)
+                  (sr-array-set! crcn j COERCION-IDENTITY-IMDT))
+                (sr-array-set! crcn c-i val-crcn)
+                (Assign (Var-id id?) FALSE-IMDT)]
+               [else UNIT-IMDT])]
+             [else
+              (assign$ c-i (op$ + (op$ - i TYPE-TUPLE-ELEMENTS-OFFSET)
+                                COERCION-TUPLE-ELEMENTS-OFFSET))
+              (sr-array-set! crcn c-i val-crcn)]))
+          (cond$
+           [id? COERCION-IDENTITY-IMDT]
+           [else
+            (assign$ iters (op$ + t1-count TYPE-TUPLE-ELEMENTS-OFFSET))
+            (assign$ i-init (op$ + t2-count TYPE-TUPLE-ELEMENTS-OFFSET))
+            (repeat$ (i i-init iters) (_ UNIT-IMDT)
+              (assign$ t1a (sr-tagged-array-ref t1 TYPE-TUPLE-TAG i))
+              (assign$ c-i (op$ + (op$ - i TYPE-TUPLE-ELEMENTS-OFFSET)
+                                COERCION-TUPLE-ELEMENTS-OFFSET))
+              (sr-array-set! crcn c-i COERCION-IDENTITY-IMDT))
+            (sr-tag-value crcn COERCION-MEDIATING-TAG)])]
+         [else
+          ;; already handled in make_coercion, I will leave it for completeness
+          (sr-alloc "failed-coercion" COERCION-FAILED-TAG `(("label" . ,l)))])))
```

Comment:
A few things:
1) This needs to be better documented
2) The cases for (= t1 t2) and (> t1 t2) could be the same case. ie (>= t1 t2). One could claim that having two cases is more efficient for (= t1 t2), but I don't think we should be overly concerned with optimizing this case. Particularly when the other case shouldn't be much slower.
3) The failed coercion at the end should just print an error message and exit instead of failing silently.

Reply:
1. Do you have suggestions?
2. I already went the extra mile and optimized it, I do not think there is any reason to tangle both cases again.
3. Do you mean we should get rid of the failure coercion everywhere? Also, this branch does not fail silently, it creates and failure coercion to be handled by compose/the cast interpreter later. It also will never get executed anyway.

Reply:
1. How about a block of text at the beginning that explains what this code is doing, any invariants it relies on (such as the fact that subtyping is getting applied here), and why you made the design choices that you made. In general it is also nice to leave comments on any piece of code that wasn't immediately obvious. In this case I know we had some trouble understanding what the code was suppose to be doing the first time around.
2. Do we actually know that it is an optimization? If the other case comes up frequently it could be slower. If you want to keep it make sure that there is a switch that allows us to check the benefit, by enabling and disabling the optimization. Also "tangling" the two cases isn't really an issue because the first is a sub-case of the second already. The tradeoff of this optimization is that there is more code to understand, debug, and keep up to date in the future. 
3. If the coercion is in a higher order context there is a chance that the else branch will execute without halting the program, because the failure coercions is only a sub coercion of a larger coercion.
Even if the failed coercion does get triggered it would be nicer if we had an error message that said precisely what we know about this line. IE where it is, and that we think this shouldn't happen because ...

Reply:
1. Done.
2. I disagree with your comment because it is a trade-off that depends on the input code and you have to choose one way to do it. Anyway, I got rid of the branching.
3. The else branch will never execute anyway because it is taken care of by the `make-coercion` function, I got rid of it.